platform/ios/src/MGLMapView.mm

```diff
@@ -2341,6 +2346,88 @@ - (void)setShowsScale:(BOOL)showsScale
+    
+    NSString *layerIdentifier = @"extrudedBuildings";
+    if (showsBuildings) {
+        MGLSource *source = [self.style sourceWithIdentifier:@"composite"];
+        if (source) {
+            
+            MGLFillExtrusionStyleLayer* layer = [[MGLFillExtrusionStyleLayer alloc] initWithIdentifier:layerIdentifier source:source];
+            layer.sourceLayerIdentifier = @"building";
+            layer.predicate = [NSPredicate predicateWithFormat:@"extrude == 'true' AND CAST(height, 'NSNumber') > 0"];
+            layer.fillExtrusionBase = [NSExpression expressionForKeyPath:@"min_height"];
+            layer.fillExtrusionHeight = [NSExpression expressionForKeyPath:@"height"];
+            
+            // Set the fill color to that of the existing building footprint layer, if it exists.
+            MGLFillStyleLayer* buildingLayer = (MGLFillStyleLayer*)[self.style layerWithIdentifier:@"building"];
+            if (buildingLayer) {
+                if (buildingLayer.fillColor) {
+                    layer.fillExtrusionColor = buildingLayer.fillColor;
+                } else {
+                    layer.fillExtrusionColor = [NSExpression expressionForConstantValue:[UIColor whiteColor]];
+                }
+                
+                layer.fillExtrusionOpacity = [NSExpression expressionForConstantValue:@0.75];
+            }
+            
+            MGLStyleLayer* labelLayer = [self.style layerWithIdentifier:@"waterway-label"];
+            if (labelLayer) {
+                [self.style insertLayer:layer belowLayer:labelLayer];
+            } else {
+                [self.style addLayer:layer];
+            }
+        }
+    } else {
+        MGLStyleLayer *layer = [self.style layerWithIdentifier:layerIdentifier];
+        if (layer) [self.style removeLayer:layer];
+    }
+    _showsBuildings = showsBuildings;
+}
+
+- (BOOL)showsTraffic {
+    NSString *sourceIdentifier = @"mapbox://mapbox.mapbox-traffic-v1";
+
+    for (MGLStyleLayer *layer in self.style.layers) {
+        if ([layer isKindOfClass:[MGLForegroundStyleLayer class]]) {
+            MGLForegroundStyleLayer *trafficLayer = (MGLForegroundStyleLayer *)layer;
+            if ([trafficLayer.sourceIdentifier isEqualToString:sourceIdentifier]) {
+                _showsTraffic = layer.isVisible;
+                return _showsTraffic;
+            }
+            
+        }
+    }
+    _showsTraffic = NO;
+    return _showsTraffic;
+}
+
+- (void)setShowsTraffic:(BOOL)showsTraffic
+{
+    _showsTraffic = showsTraffic;
+    NSString *sourceIdentifier = @"mapbox://mapbox.mapbox-traffic-v1";
+    for (MGLStyleLayer *layer in self.style.layers) {
+        if ([layer isKindOfClass:[MGLForegroundStyleLayer class]]) {
+            MGLForegroundStyleLayer *trafficLayer = (MGLForegroundStyleLayer *)layer;
+            if ([trafficLayer.sourceIdentifier isEqualToString:sourceIdentifier]) {
+                [layer setVisible:showsTraffic];
```

Comment:
Unlike `showsBuildings`, this setter only sets the visibility of preexisting traffic layers rather than creating new layers based on the style. Admittedly, the latter is difficult to do in a reliable way. Regardless, these properties need to have consistent behavior.

Reply:
The [traffic plugin](https://github.com/mapbox/mapbox-plugins-ios/blob/master/TrafficPlugin/MBXTrafficPlugin.m) creates new layers for traffic congestion lines, though the colors are hardcoded and aren’t guaranteed to match the style well. Maybe that’s sufficient for this feature, but if not, we’d need to design heuristics for determining the colors and line widths that would look good atop the roads in the style. I suppose the same problem exists to a lesser extent with showing POIs on a style (such as Satellite) that lacks them.

Alternatively, this feature can be exclusively about toggling layers that are already part of the style. But in that case, we might need to make an exception for 3D buildings, because the main Mapbox-designed styles lack 3D buildings in the first place.

Reply:
I'm following along the lines of showing the default style for traffic. I think providing an option to personalize the style would be better but I don't think it's a scope for this PR.